### ISSUE TITLE
Slow restart of lizardfs services in large setups

### ISSUE THREAD

**Issue by r5r3** [lizardfs/lizardfs#645]
We have a rather large installation of LizardFS in use (about 200M files). We always have a problem when installing updates. Restarting the services takes too long. This applies in particular to the master and the chunk servers. An update results in a downtime of at least one hour. 

Is there any way to speed up the restart? For example, not to execute checks directly at the start of the services, but asynchronously later? Is something like that conceivable?

The number of files in the system will increase rather than decrease in the future, so the problem will unfortunately get worse.

**Comment by guestisp** [comment]
It could be related to libJudy.
See #587

**Comment by psarna** [comment]
It's not usually the matter of checks, but with loading all metadata from disk.

Things you could do:
- Check if your data directory (/var/lib/mfs or /var/lib/lizardfs) is mounted on an SSD drive. If not - consider it, it will speed up startup considerably.
- Indeed, check if your build links to libJudy, it's better if it does.
- Clear your trash. Millions of files in trash can slow down the process, especially if you have a libJudy-less installation.
- You could install perf, run it on starting master (something like `perf top -p $(pgrep mfsmaster)` should suffice) and share the results here. It gives a lot of context as what needs to be optimized.

**Comment by r5r3** [comment]
Thank you for the suggestions!
We use your official rpms for RedHat, which means that the master links to libJudy. 

The metadata is stored on a ZFS-mirror made from SSDs (connected via SAS3). The metadata file is currently about 21 GB. Reading it takes about 20s. That is actually somewhat below the expected speed of the disks. However, the master is not starting up within less than a minute (which would be perfectly fine). 

Trash: there are only a few thousand files in trash, that should not hurt. 

Perf: yes, I will provide information collected with `perf` when I have to restart the master the next time. Unfortunately, I can not do that directly as the restart takes too long and too many users would be affected. 

Chunkservers: What about the chunk-servers. They spend a lot of time scanning the disks at startup. Would it be possible to directly start the operation and scan the disks later? The assumption that the content of the disks is not changed when the service is restarted seems to be reasonable.

**Comment by 4Dolio** [comment]
I have similarly large metadata, but apparently even slower disks.  Master starts can take up to 20+minutes and so our HA is critical for zero downtime for routine maintainence.  I can not lookup the recomended upgrade process for LFS. I would hope, and if not we should work on, the ability to upgrade LFS with zero downtime by: upgrade CS first. Upgrade Shadows next. Optionally upgrade clients, master & CS need to be able to be able to always backwards compatible downgrade communications with CS, Masters, Clients.  Finally. Execute a HA transition to an upgraded shadow, last upgrade prrvious master.

**Comment by 4Dolio** [comment]
Yup. That will work...

**Comment by r5r3** [comment]
Thank you for the suggestion of an update order which will reduce the downtime. 

However, in our case it will not be reduced to zero. A part of our chunkservers run on redundant hardware (JBOD connected to two servers). For these chunkservers we do not use replication (would be too expensive). That means, an update will at least cause the downtime needed for a restart of chunkservers. Due to the initial disk scanning, this restart is extremely slow. The largest chunkservers operate on 160TB ZFS pools (currently filled with about 50TB of chunks). The last restart of such a server took about 66min. 

I understand that (for the chunkservers) the actual reason of the problem is our decision not to buy the disks twice. But I think that there is some room for shortening the startup time. One way would be to postpone the initial scanning of the disks.

**Comment by guestisp** [comment]
Chunkservers won't be able to serve chunks if initial scanning is not conpleted.if you don't have redundant chunks, even after postponing the initial scan, you won't solve the issue, you'll end up with and empty chunkserver unable to serve it's chunks until the end of scan.

Your goal is to have a fast daemon startup or a fast daemon serving it's own chunks?

**Comment by 4Dolio** [comment]
It simply can not be done with data at a goal of 1.  Frankly, you might as well put the data directly on the zfs and not bother with lizardfs (almost).  If you are not above 50% use, you could go to goal=2 and then you can do it with zero downtime.  Each LizardFS component has a non zero boot/initialize time, so the pieces need to have redundant peers in order to leverage the advantages LFS provides.

**Comment by r5r3** [comment]
The comment of @4Dolio could basically be summarized to "LizardFS is not suitable for large setups => use ZFS". I guess that the people from Skytechnology have a somewhat different opinion, and so do I.

We use LizardFS because it offers the possibility to combine storage servers with and without redundant hardware into a single namespace. We use replication goals for non-redundant hardware only. We would have about ten different mount points for one purpose if we would use ZFS only. That would be much less convenient for the users (and administrators as well). In addition, the easier migration to new server hardware makes it worth using a distributed system. Before we made our decision to use LizardFS, we compared it to all other open source distributed filesystems. LizardFS has not the best performance (Lustre was much better), but it did have the best set of features and flexibility. 

In HPC environments, it is quite common not to do any data replication. The mean reason is that buying the hardware twice for petabyte-scale systems is not affordable. 

It is totally clear that (by design) restarting not replicated services is not possible without downtime. But, the question is how many downtime is really necessary. To come back to @guestisp's question: The goal would be to have a fast daemon serving it's own chunks. I could imagine, that the startup time could be significantly reduced when the chunkserver would maintain a list of chunks in a file. This file could be written when the service is stopped and read on startup (similar to the behavior of the master). I could imagine that faster startup times would also be of interest for other users who use replication. During the long startups (can obviously be hours) their files would always be undergoal.

**Comment by Zorlin** [comment]
@r5r3 I did not interpret it that way at all. Large setups with 1 copy are not (IMO) suitable for LizardFS if the availability of chunks is necessary. Large setups with erasure coding or plain more-than-1-copy setups are perfectly suitable to LizardFS and @4Dolio did not say otherwise.

**Comment by Blackpaw** [comment]
>LizardFS is not suitable for large setups => use ZFS

Disagree most strongly. I'm amazed anyone creates huge distributed volumes without redundancy. If a node goes down, you're toast.

**Comment by r5r3** [comment]
> I'm amazed anyone creates huge distributed volumes without redundancy.

Every large computation center with Lustre is doing exactly that. Reliability is ensured by having two servers serving the same disks. Switching a ZFS pool from one machine to another is only a matter of seconds.

> If a node goes down, you're toast.

That is why we do backups on tape which is by orders of magnitude cheaper then hard disks. 

> Large setups with erasure coding or plain more-than-1-copy setups are perfectly suitable to LizardFS

Beside of the expensive disk space, this is also a question of acceptable network load. 

Sorry, but I did not intend to start a discussion about the sense of data replication. There is no alternative for important, not easily reproducible data, no question! In some situations, however, it is simply a trade-off between the cost of storage and the cost of computing time (to reproduce the data). And that is the reason why I was questioning whether or not the current startup procedure of the LizardFS services is already optimal.  And, as mentioned above, a reduced undergoal time should also be something users with data replication are interested in.

**Comment by 4Dolio** [comment]
I did say "almost" and did not intend to imply there were no benefits from LFS even with goal=1.  As it is now, you will have downtime, and scans will be slow... you could speed this up if you can parallize your CS.. ie: use many individual disks rather than a zfs pool composed of many disks, but you will loose any RaidZ that you have from zfs (i assume you have disk redundency).

Your idea of dumping Chunk file data to disk upon stop of CS and loading it on start may have some merit, but the coding of such a feature is above my paygrade.  I do feel your pain, when we do a full cluster stop it takes our CS about a half our to come back online.  I can totally see it taking hours given more data.

**Comment by pkali** [comment]
I'd say this is a valuable idea worth considering!

**Comment by AnsgarKlein** [comment]
> Every large computation center with Lustre is doing exactly that. Reliability is ensured by having two servers serving the same disks.

I'm actually in the process of planning a similar storage system (~0.8-1.4 PB) and still deciding which software to use. SAS hardware with redundant hosts are the only way I can achieve this kind of storage capacity with the available budget. I guess having no replication saves network bandwidth as well (?).
I'm currently leaning towards LizardFS but having a restart time of an hour would make me quite nervous.

> the startup time could be significantly reduced when the chunkserver would maintain a list of chunks in a file. This file could be written when the service is stopped and read on startup (similar to the behavior of the master).

Wouldn't something like this be possible?

**Comment by 4Dolio** [comment]
Without replication above goal=1 you can never do upgrades or reboot a chunk server without having downtime.  Not until a new CS fast start scan cache feature is added that is...

> I guess having no replication saves network bandwidth as well (?).

Local bandwidth is essentially free, so long as clients can still get work done and you never saturate for the sake of replication.

**Comment by r5r3** [comment]
@4Dolio You suggested to parallelize the CS to speedup restart. You are right about the z-raid. We definitively want to keep that, but our pools are really large with 30 disks each (three vdevs per pool). We could easily split that up into smaller pools. (By the way, in doing so, we would leverage another benefit of LizardFS that is available with goal=1: migration without interruption of service). To get a parallel initial scanning of the disks, would it be necessary to have multiple instances of the chunk-server or is it sufficient to present multiple zpools as multiple disks to one instance of a chunk-server? 

@pkali Thanks for adding the "enhancement" label!

**Comment by 4Dolio** [comment]
Back to basics, a CS like ZFS likes to control the disks, sort of. You don't have to, but a CS with many disks accesses them in parallel.  At some point you still encounter underlying disk channel/io limits.  But when you put zfs or even hw raid inbetween your CS and the disks you lose this parallism, and hope hope that there is some benefit from the tradeoff.. our direct disk jbod server is 2 to 3 times faster to scan disks than a similar capacity hw raid CS.  Im not sure what the scan does exactly but perhaps zfs could be tuned with a larger arc and be made to prefetch chunk file metadata, but scanning may also be reading headers from each file?  With a single goal you are really back to being at the mercy of each single disk and all the additional overhead you add to the mix...

Having goals above 1 should not be thought of as losing space buy a increasing disk speeds as well as redundancy.

**Comment by r5r3** [comment]
> But when you put zfs or even hw raid inbetween your CS and the disks you lose this parallism

Partially true, zfs can read and write to each vdev in a pool in parallel. That is one of the reasons why hw raid is usually slower than jbod+zfs (tested with areca hw raid).

> perhaps zfs could be tuned with a larger arc and be made to prefetch chunk file metadata

We have indeed a very fast ARC2 (1TB for each pool, max 1.7GB/s). I could imagine that more prefetching could help a little bit. However, it would be necessary to know what the CS is actually doing during its scan procedure. If only meta data is read, than it would possibly help to convert the ARC2 devices to meta-devices (ZFS supports separate storage of meta-data).

> With a single goal you are really back to being at the mercy of each single disk

We are definitively not thinking about using single disks. With erasure coding, that would lead to the same capacity than ZFS with raidz2, but LizardFS-EC is much slower than ZFS and drives up network traffic (with LizardFS, the clients are actually doing the EC).

> Having goals above 1 should not be thought of as losing space buy a increasing disk speeds as well as redundancy.

It turns out that goals above one have higher latencies when dialing with very small files. But the main reason for using goal 1 is the available amount of money. 

However, I think my question was misunderstood. I only wanted to know if the CS is scanning all disks defined in the HDD file in parallel or one a a time. But I can of course try by myself.

**Comment by 4Dolio** [comment]
Yes. CS scan all disks in parallel.  Sorry for the complex answer before. All good counterpoints.

So, having multiple zfs 'disk' may help. I would guess.
